File: codesphere/frontend/src/lib/languages.ts
/** The set of languages students may submit in - mirrors
 * backend/app/core/languages.py. A curated "popular contest languages"
 * set (5), not Judge0's full 60+ catalog. */

export type LanguageId = 'c' | 'cpp' | 'java' | 'python3' | 'javascript'

export interface LanguageInfo {
  id: LanguageId
  label: string
  /** Monaco's built-in language id for syntax highlighting - NOT always
   * the same as `id` (Python's Monaco id is "python", not "python3"). */
  monacoId: string
  defaultTemplate: string
}

export const LANGUAGES: Record<LanguageId, LanguageInfo> = {
  c: {
    id: 'c',
    label: 'C',
    monacoId: 'c',
    defaultTemplate: '#include <stdio.h>\n\nint main() {\n    \n    return 0;\n}\n',
  },
  cpp: {
    id: 'cpp',
    label: 'C++',
    monacoId: 'cpp',
    defaultTemplate: '#include <iostream>\nusing namespace std;\n\nint main() {\n    \n    return 0;\n}\n',
  },
  java: {
    id: 'java',
    label: 'Java',
    monacoId: 'java',
    // Judge0 writes the source to Main.java before compiling - the public
    // class must be named exactly "Main" or every submission fails to
    // compile regardless of the student's actual logic.
    defaultTemplate: 'import java.util.Scanner;\n\npublic class Main {\n    public static void main(String[] args) {\n        Scanner sc = new Scanner(System.in);\n        \n    }\n}\n',
  },
  python3: {
    id: 'python3',
    label: 'Python 3',
    monacoId: 'python',
    defaultTemplate: '',
  },
  javascript: {
    id: 'javascript',
    label: 'JavaScript (Node.js)',
    monacoId: 'javascript',
    defaultTemplate: '',
  },
}

export const DEFAULT_LANGUAGE: LanguageId = 'c'

export const LANGUAGE_LIST: LanguageInfo[] = Object.values(LANGUAGES)

File: codesphere/backend/app/core/languages.py
"""The set of languages students may submit in.

A curated "popular contest languages" set (5), not the full 60+ language
catalog Judge0 CE exposes - keeps the editor dropdown, per-language
starter templates, and testing surface manageable for a college coding
contest. `judge0_id` values are Judge0 CE's stable, long-standing IDs -
matching the public `ce.judge0.com` demo instance this app defaults to.

JudgeService/SyncJudgeService (app/services/judge_service.py) already
accept an optional `language_id` constructor argument - this registry is
just the lookup table every call site uses to pick the right one, instead
of always falling back to settings.judge0_c_language_id (C).
"""

from enum import Enum

from pydantic import BaseModel


class SupportedLanguage(str, Enum):
    C = "c"
    CPP = "cpp"
    JAVA = "java"
    PYTHON3 = "python3"
    JAVASCRIPT = "javascript"


class LanguageInfo(BaseModel):
    id: SupportedLanguage
    judge0_id: int
    display_name: str
    # Monaco's built-in language id for syntax highlighting - NOT always
    # the same as `id` (Python's Monaco id is "python", not "python3").
    monaco_id: str


LANGUAGES: dict[SupportedLanguage, LanguageInfo] = {
    SupportedLanguage.C: LanguageInfo(
        id=SupportedLanguage.C, judge0_id=50, display_name="C", monaco_id="c"
    ),
    SupportedLanguage.CPP: LanguageInfo(
        id=SupportedLanguage.CPP, judge0_id=54, display_name="C++", monaco_id="cpp"
    ),
    SupportedLanguage.JAVA: LanguageInfo(
        id=SupportedLanguage.JAVA, judge0_id=62, display_name="Java", monaco_id="java"
    ),
    SupportedLanguage.PYTHON3: LanguageInfo(
        id=SupportedLanguage.PYTHON3, judge0_id=71, display_name="Python 3", monaco_id="python"
    ),
    SupportedLanguage.JAVASCRIPT: LanguageInfo(
        id=SupportedLanguage.JAVASCRIPT,
        judge0_id=63,
        display_name="JavaScript (Node.js)",
        monaco_id="javascript",
    ),
}

DEFAULT_LANGUAGE = SupportedLanguage.C
